<!DOCTYPE html>
<html>

  <head>
    <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1">

<link rel="stylesheet" href="/css/main.css">

<!-- Custom stylesheets -->


    <!-- Begin Jekyll SEO tag v2.7.1 -->
<title>How to Obtain Cropped Nucleotide Sequences and More | Cheuk’s Blog</title>
<meta name="generator" content="Jekyll v4.2.1" />
<meta property="og:title" content="How to Obtain Cropped Nucleotide Sequences and More" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Using BioPerl, get subsets of a nucleotid sequence." />
<meta property="og:description" content="Using BioPerl, get subsets of a nucleotid sequence." />
<meta property="og:site_name" content="Cheuk’s Blog" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2017-08-14T09:43:00-07:00" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="How to Obtain Cropped Nucleotide Sequences and More" />
<script type="application/ld+json">
{"headline":"How to Obtain Cropped Nucleotide Sequences and More","dateModified":"2017-08-14T09:43:00-07:00","datePublished":"2017-08-14T09:43:00-07:00","mainEntityOfPage":{"@type":"WebPage","@id":"/bioperl/2017/08/14/howto-bioperl.html"},"description":"Using BioPerl, get subsets of a nucleotid sequence.","@type":"BlogPosting","url":"/bioperl/2017/08/14/howto-bioperl.html","@context":"https://schema.org"}</script>
<!-- End Jekyll SEO tag -->


    
  <head>

  <body>
    <div class="wrapper">
      <nav>
  <a href="/" class="nav-title">
    Cheuk's Blog
  </a>

  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  <a class="nav-links" href="/resources/">Resources</a>
  
  
  
  
  
  
  
  
  
  
  
  
  
  <a class="nav-links" href="/about/">About</a>
  
  
</nav>

      
      <article class="post">
  <h1 class="title">How to Obtain Cropped Nucleotide Sequences and More</h1>
  <span>
    <span>
    Tags:
    
    <a href="/tag/bioperl">
      bioperl
    </a>
    
    <a href="/tag/perl">
      perl
    </a>
    
    <a href="/tag/bioinformatics">
      bioinformatics
    </a>
    
    </span>
    <span class="meta">14 Aug 2017</span>
  </span>

  <p>In my opinion, there isn’t much documentation on the BioPerl site: just vague
examples and whatnot.</p>

<p>Say for instance you want to get the entire nucleotide sequence of a particular
organism, but it must be split up because you only want to deal with the
proteins, and not the entire organism. So less like this:</p>

<pre><code class="language-fasta">&gt;NC_002695.1 Escherichia coli O157:H7 str. Sakai, complete genome
AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTCTCTGACAGCAGC
TTCTGAACTGGTTACCTGCCGTGAGTAAATTAAAATTTTATTGACTTAGGTCACTAAATACTTTAACCAA
TATAGGCATAGCGCACAGACAGATAAAAATTACAGAGTACACAACATCCATGAAACGCATTAGCACCACC
ATTACCACCACCATCACCACCACCATCACCATTACCATTACCACAGGTAACGGTGCGGGCTGACGCGTAC
AGGAAACACAGAAAAAAGCCCGCACCTGACAGTGCGGGCTTTTTTTTCGACCAAAGGTAACGAGGTAACA
ACCATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAACGTTTTCTGCGGGTTGCCGATA
TTCTGGAAAGCAATGCCAGGCAGGGGCAGGTGGCCACCGTCCTCTCTGCCCCCGCCAAAATCACCAACCA
CCTGGTGGCGATGATTGAAAAAACCATTAGCGGCCAGGATGCTTTACCCAATATCAGCGATGCCGAACGT
ATTTTTGCCGAACTTCTGACGGGACTCGCCGCCGCCCAGCCGGGATTCCCGCTGGCGCAATTGAAAACTT
TCGTCGACCAGGAATTTGCCCAAATAAAACATGTCCTGCATGGCATTAGTTTGTTAGGGCAGTGCCCGGA
TAGCATTAACGCTGCGCTGATTTGCCGTGGCGAGAAAATGTCGATCGCCATTATGGCCGGCGTATTAGAA
GCGCGCGGTCACAACGTTACCGTTATCGATCCGGTCGAAAAACTGCTGGCAGTGGGGCATTACCTCGAAT
CTACTGTCGATATTGCAGAGTCCACCCGCCGTATTGCGGCAAGTCGTATTCCGGCTGATCACATGGTGCT
GATGGCAGGTTTCACCGCCGGTAATGAAAAAGGCGAACTGGTGGTACTTGGACGCAACGGTTCCGACTAC
TCCGCGGCGGTGCTGGCTGCCTGTTTACGCGCCGATTGTTGCGAGATTTGGACGGACGTTGACGGGGTAT
ATACCTGCGACCCGCGTCAGGTGCCCGATGCGAGGTTGTTGAAATCGATGTCCTACCAGGAAGCGATGGA
GCTTTCCTACTTCGGCGCTAAAGTTCTTCACCCCCGCACCATTACCCCCATCGCCCAGTTCCAGATCCCT
TGCCTGATTAAAAATACCGGAAATCCTCAAGCTCCAGGTACGCTCATTGGTGCCAGTCGTGATGAAGACG

...
</code></pre>

<p>And more like this:</p>

<pre><code class="language-fasta">&gt;SDY_PA01
ATGTCTGAATTAGTTGTGTTTAAAGCAAATGAATTAGCAGTAAGCCGTTATGATCTAACTGAACATGAAA
CCAAGCTAATTCTGTTTTGCGTTGCAAAGTTGAACCCTACAATTGAAAACCCAACAAGGGATGAATTAAC
AGTTAAATTCTCATGTTCTGAATATGCACGCACTATGGGCTTAAGTTATGAAAATGCTTGGGGAAGATTG
AACAGTGCAACTAGTGATTTATTTAAACGCTCTGTTGAATTAATTTATCCTACAGGAGCAGTTTCTAAAC
GAATTTTCAATTGGACGGAATACGCTGAGTTCAATAGAGAAGAACAAACTGTAACATTAGTATTTAGCTC
ATATATTCAACCACTTTTATTTCATTTAAAAAAATTCATTAAATACAACCTTGAGCATGTTAAAGCCTTT
GAAAATAAATACTCAATGCGAATATATGAATGGTTACTAAAGGAGCTTTCACAAAGAAAAACGCATAGGG
GAAACATTGAGATAAGTATCAAGGAATTTAAGTTTATGCTCATGCTTGAAAAAAACTACCCTTTATATGC
AGAATTGAATCGCTGGATCCTAAAACCAGTTACAAATGACTTAAACACTTACAGCAATATGAAGTTGACT
ATTGAAAAACGCGGTCGTCCTGCTGACACACTGATCTTTCAAGTTGAACTGGATAAACAAATTGACCTTG
TGACTGAACTAGCAAAAGATCCCGCATCAAAAAAAGAAGATAAGACAATCCGTTTAACGCCTGAAAATCG
TCTTCATGAGGGGCTAAAAACAACATTGCATGATGCTTTAACTGCAAAAATTCAACTGACTAGTTTTGAA
GCAAAATTCCTGAGCGATATGCAAAGCAAGTACGATCTTAACGGCTCATTTACATGGCTGACTCAAAAGC
AACGAACTACGCTAGAGAAAATTCTGGCTAAATACGGGCGGATATGA
&gt;SDY_PA02
GTGGCTTCTGTTTCTATCAACTGTCCCTCCTGTTCAGCTACTGACGGGGGGGTGCGTAACGGCAAAAGCA
CTGCCGGACATCAGCGCTATCTCTGCTCTCACTGCCGTAAAACATGGCAACTGCAGTTCACTTACACCGC
TTCTCAACCCGGTACGCACCAGAAAATCATTGATATGGCCATGAATGGCGTTGGATGCCGGGCAACTGCA
CGCATTATGGGCGTTAGCCTCAACACGATTTTACGTCACTTAAAAAACTCAGGCCGCAGTCGGTAA
&gt;SDY_PA03
ATGGACGAACAGTGGGGATACGTCGGGGCTAAATCGCGCCAGCGCTGGCTGTTTTACGCGTATGACAGGC
TCCGGAAGACGGTTGTTGCGCACGTATTCGGTGAACGCACTATGGCGACGCTGGGGCGTCTTATGAGCCT
GCTGTCCCCCTTTGACGTGGTGATATGGATGACGGATGGCTGGCCGCTGTATGAATCCCGCCTGAAGGGA
AAGCTGCACGTAATCAGCAAGCGATATACGCAGCGAATTGAGCGGCATAACCTGAATCTGAGGCAGCACC
TGGCACGGCTGGGACGGAAGTCGCTGTCGTTCTCAAAATCGGTGGAGCTGCATGACAAAGTCATCGGGCA
TTATCTGAACATAAAACACTATCAATAA
</code></pre>

<p>As you may or may not notice, these are 2 completely different organisms, but
you get the drift.</p>

<h2 id="pre-requisites">Pre-requisites</h2>

<p>BioPerl must be installed, along with Perl.</p>

<h2 id="disclaimers">Disclaimer(s)</h2>

<p>I don’t do Perl. I don’t like it that much, and I’m not very experienced in it.
I also don’t do Biology.</p>

<h2 id="the-code">The Code</h2>

<p>Start your file with the basics:</p>

<div class="language-perl highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#!/usr/bin/env perl</span>
<span class="k">use</span> <span class="nv">strict</span><span class="p">;</span>
<span class="k">use</span> <span class="nv">warnings</span><span class="p">;</span>

<span class="c1"># We'll be getting nucleotides from GenBank database (NCBI)</span>
<span class="k">use</span> <span class="nn">Bio::DB::</span><span class="nv">GenBank</span><span class="p">;</span>
<span class="k">use</span> <span class="nn">Bio::DB::Query::</span><span class="nv">GenBank</span><span class="p">;</span>
</code></pre></div></div>

<p>To specify the specific sequence you want to get, BioPerl recommends creating
a query object.</p>

<div class="language-perl highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">my</span> <span class="nv">$query</span> <span class="o">=</span> <span class="p">"</span><span class="s2">NC_002695 [ACCN] OR NC_002127 [ACCN] OR NC_002128 [ACCN]</span><span class="p">";</span>
<span class="k">my</span> <span class="nv">$query_obj</span> <span class="o">=</span> <span class="nn">Bio::</span><span class="nv">DB:Query::GenBank</span><span class="o">-&gt;</span><span class="k">new</span><span class="p">(</span><span class="o">-</span><span class="s">db</span> <span class="o">=&gt;</span>    <span class="p">'</span><span class="s1">nucleotide</span><span class="p">',</span>
                                            <span class="o">-</span><span class="s">query</span> <span class="o">=&gt;</span> <span class="nv">$query</span><span class="p">);</span>
</code></pre></div></div>

<p>I will be using accesion numbers to get the specific sequences. Note that the
binary operators <code class="language-plaintext highlighter-rouge">OR</code> (other binary operators include <code class="language-plaintext highlighter-rouge">AND</code>) must be in
uppercase. For a complete list of query fields, see <a href="https://www.ncbi.nlm.nih.gov/entrez/query/static/help/Summary_Matrices.html#Search_Fields_and_Qualifiers">this</a>.</p>

<div class="language-perl highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Start querying</span>
<span class="k">my</span> <span class="nv">$gb</span> <span class="o">=</span> <span class="nn">Bio::</span><span class="nv">DB:GenBank</span><span class="o">-&gt;</span><span class="k">new</span><span class="p">;</span>
<span class="k">my</span> <span class="nv">$seq_objs</span> <span class="o">=</span> <span class="nv">$gb</span><span class="o">-&gt;</span><span class="nv">get_Stream_by_query</span><span class="p">(</span><span class="nv">$query_obj</span><span class="p">);</span>

<span class="c1"># Iterate through all the sequences gotten</span>
<span class="k">while</span> <span class="p">(</span><span class="k">my</span> <span class="nv">$seq_obj</span> <span class="o">=</span> <span class="nv">$seq_objs</span><span class="o">-&gt;</span><span class="nv">next_seq</span><span class="p">)</span> <span class="p">{</span>
    <span class="c1"># Iterate through all the features</span>
    <span class="k">for</span> <span class="k">my</span> <span class="nv">$feat_obj</span> <span class="p">(</span><span class="nv">$seq_obj</span><span class="o">-&gt;</span><span class="nv">get_SeqFeatures</span><span class="p">)</span> <span class="p">{</span>
        <span class="c1"># Only print features with primary tag "CDS", "protein_id", and</span>
        <span class="c1"># "locus_tag" (I'm not sure what "CDS" stands for)</span>
        <span class="k">if</span> <span class="p">(</span><span class="nv">$feat_obj</span><span class="o">-&gt;</span><span class="nv">primary_tag</span> <span class="ow">eq</span> <span class="p">"</span><span class="s2">CDS</span><span class="p">"</span> <span class="ow">and</span>
            <span class="nv">$feat_obj</span><span class="o">-&gt;</span><span class="nv">has_tag</span><span class="p">("</span><span class="s2">protein_id</span><span class="p">")</span> <span class="ow">and</span>
            <span class="nv">$feat_obj</span><span class="o">-&gt;</span><span class="nv">has_tag</span><span class="p">("</span><span class="s2">locus_tag</span><span class="p">"))</span> <span class="p">{</span>
                <span class="k">print</span> <span class="p">"</span><span class="s2">&gt;</span><span class="p">",</span> <span class="nv">$feat_obj</span><span class="o">-&gt;</span><span class="nv">get_tag_values</span><span class="p">("</span><span class="s2">locus_tag</span><span class="p">"),</span> <span class="p">"</span><span class="se">\n</span><span class="p">";</span>
                <span class="k">print</span> <span class="nv">$feat_obj</span><span class="o">-&gt;</span><span class="nv">spliced_seq</span><span class="o">-&gt;</span><span class="nv">seq</span><span class="p">,</span> <span class="p">"</span><span class="se">\n</span><span class="p">";</span>
        <span class="p">}</span>
    <span class="p">}</span>
<span class="p">}</span>
</code></pre></div></div>

<p>To explain the above code a bit more in depth, here is the raw text of a
GenBank file, copied straight off of the NCBI database, features only:</p>

<pre><code class="language-genbank">FEATURES             Location/Qualifiers
     source          1..5498450
                     /organism="Escherichia coli O157:H7 str. Sakai"
                     /mol_type="genomic DNA"
                     /strain="Sakai"
                     /sub_strain="RIMD 0509952"
                     /serovar="O157:H7"
                     /db_xref="taxon:386585"
     misc_feature    1..5498450
                     /note="REFSEQ gene predictions performed by GeneMark
                     2.4/GeneMark.hmm 2.0 with comparison to original submitter
                     provided annotation."
     gene            190..273
                     /gene="thrL"
                     /locus_tag="ECs0001"
                     /db_xref="GeneID:913387"
     CDS             190..273
                     /gene="thrL"
                     /locus_tag="ECs0001"
                     /note="involved in threonine biosynthesis; controls the
                     expression of the thrLABC operon"
                     /codon_start=1
                     /transl_table=11
                     /product="thr operon leader peptide"
                     /protein_id="NP_308028.1"
                     /db_xref="GeneID:913387"
     gene            354..2816
                     /gene="thrA"
                     /locus_tag="ECs0002"
                     /db_xref="GeneID:913388"
</code></pre>

<p>There seems to be only one method of distinguishing genes that are proteins
from genes that have nothing to do with the proteins: in the CDS, there is one
tag “protein_id” that only proteins have (it corresponds to the protein’s
protein product). Genes that aren’t proteins don’t have this tag. Using this
“distinguishing” feature, I was able to filter out all the genes that aren’t
proteins.</p>

<p>Our research heavily used the locus tag of proteins, but not the protein
product as much.</p>


</article>

<footer class="post-foot">
  <span class="prev">
  
    <span>Prev</span>
    <a href="/writing/2017/06/28/cw-fire.html">[Writing] Fire</a>
  
  </span>
  <span class="next">
  
    <span>Next</span>
    <a href="/tutorial/2017/12/19/googletest-tutorial.html">Google Test Basic Tutorial</a>
  
  </span>
</footer>

    </div>
  </body>

</html>
